packages/mui-utils/src/resolveProps/resolveProps.ts

```diff
@@ -14,36 +16,51 @@ export default function resolveProps<
 >(defaultProps: T, props: T) {
   const output = { ...props };
 
-  (Object.keys(defaultProps) as Array<keyof T>).forEach((propName) => {
-    if (propName.toString().match(/^(components|slots)$/)) {
-      output[propName] = {
-        ...(defaultProps[propName] as any),
-        ...(output[propName] as any),
-      };
-    } else if (propName.toString().match(/^(componentsProps|slotProps)$/)) {
-      const defaultSlotProps = (defaultProps[propName] || {}) as T[keyof T];
-      const slotProps = props[propName] as {} as T[keyof T];
-      output[propName] = {} as T[keyof T];
+  for (const key in defaultProps) {
+    if (Object.prototype.hasOwnProperty.call(defaultProps, key)) {
+      const propName = key as keyof T;
+
+      if (propName === 'components' || propName === 'slots') {
+        output[propName] = {
+          ...(defaultProps[propName] as any),
+          ...(output[propName] as any),
+        };
+      } else if (propName === 'componentsProps' || propName === 'slotProps') {
+        const defaultSlotProps = defaultProps[propName] as T[keyof T] | undefined;
+        const slotProps = props[propName] as {} as T[keyof T] | undefined;
+
+        if (!slotProps || isObjectEmpty(slotProps)) {
+          // Reduce the iteration if the slot props is empty
+          output[propName] = defaultSlotProps || ({} as T[keyof T]);
+        } else if (!defaultSlotProps || isObjectEmpty(defaultSlotProps)) {
+          // Reduce the iteration if the default slot props is empty
+          output[propName] = slotProps;
+        } else {
+          output[propName] = { ...slotProps };
 
-      if (!slotProps || !Object.keys(slotProps)) {
-        // Reduce the iteration if the slot props is empty
-        output[propName] = defaultSlotProps;
```

Comment:
`!Object.keys(anything)` is always going to be false, because it's equivalent to `![]`. I'm guessing this was meant to be `!Object.key(x).length`. Anyway, I've replaced with a simpler version.

Reply:
Did one last change, removed the empty object check optimization. Avoiding iterating an empty object isn't an interesting optimization, and it's an unlikely case, and less code is always a win.